src/Components/Components/src/ParameterCollectionExtensions.cs

```diff
@@ -41,17 +42,68 @@ public static class ParameterCollectionExtensions
                 _cachedWritersByType[targetType] = writers;
             }
 
+            // The logic is a little convoluted now that we have "Extra" parameters. We want to avoid allocations where
```

Comment:
This is interesting. I didn't expect the splatting resolution to happen here in `SetParameterProperties`. This means that splatting becomes a feature of `SetParameterProperties`, so if someone tries to write manual logic to iterate over the parameter collection (as we do in various built-in components), it won't work because they will see the actual dictionaries of parameters to be splatted.

We need to choose whether splat resolution is meant to be visible to receivers. What use cases are there for knowing what specific dictionaries were supplied, instead of just seeing the result of combining them all? If we choose only to make visible the results of combining all the dictionaries, then ideally this would be resolved at an earlier phase, either:

 * During the render tree builder itself (for each "set multiple parameters" call, we iterate over the dictionaries at that point so the output is indistinguishable from a normal series of "set attribute" calls, and to implement the overwriting logic, we scan backwards through the attribute frames and replace any with the same name). This is what I was expecting.
 * Or, inside `ParameterCollectionExtensions`, implementing the equivalent of what you have here, but in the custom iterator. Then in any typical use case you'd only observe the results of combining the dictionaries, and `SetParameterProperties` would get it automatically.

Do you think either of those would be possible?

Reply:
Update: Please ignore the above comment. Now I've read the actual code here and more of the implementation later, I see you are already doing what I was suggesting. Sorry for the confusion.